src/python/pants/engine/rules.py

```diff
@@ -427,6 +427,17 @@ def __new__(cls, union_base, union_member):
     return super().__new__(cls, union_base, union_member)
 
 
+@dataclass(frozen=True)
+class UnionMembership:
+  union_rules: Dict[type, Iterable[type]]
+
+  def is_union_member(self, union_type, tgt):
```

Comment:
Given the name `UnionMembership`, `is_union_member` is a bit redundant. If you stick with `Union` in the classname, maybe just `is_member(x, y)`... if you renamed this to something else like `Types`.